sql/core/src/main/scala/org/apache/spark/sql/execution/SizeLimitingByteArrayUnsafeRowsConverter.scala

```
@@ -0,0 +1,108 @@
+    val bos = new ByteArrayOutputStream()
+    val out = new DataOutputStream(codec.compressedOutputStream(bos))
+    // `unsafeRows.hasNext` may produce one row and buffer it, we should only call it
+    // when the limit is not hit.
+    while ((n < 0 || count < n) && unsafeRows.hasNext) {
+      val row = unsafeRows.next().asInstanceOf[UnsafeRow]
+      ensureTotalSizeIsBelowLimit(row.getSizeInBytes)
+      out.writeInt(row.getSizeInBytes)
+      row.writeToStream(out, buffer)
+      count += 1
+    }
+    out.writeInt(-1)
+    out.flush()
+    out.close()
+    Iterator((count, bos.toByteArray))
+  }
+
+  /**
+   * Decodes the byte arrays back to UnsafeRows and puts them into buffer.
+   */
+  def decodeUnsafeRows(nFields: Int, bytes: Array[Byte]): Iterator[InternalRow] = {
+    val codec = CompressionCodec.createCodec(SparkEnv.get.conf)
+    val bis = new ByteArrayInputStream(bytes)
+    val ins = new DataInputStream(codec.compressedInputStream(bis))
+
+    new Iterator[InternalRow] {
+      private var sizeOfNextRow = ins.readInt()
+
+      override def hasNext: Boolean = sizeOfNextRow >= 0
+
+      override def next(): InternalRow = {
+        ensureTotalSizeIsBelowLimit(sizeOfNextRow)
+        val bs = new Array[Byte](sizeOfNextRow)
+        ins.readFully(bs)
+        val row = new UnsafeRow(nFields)
+        row.pointTo(bs, sizeOfNextRow)
+        sizeOfNextRow = ins.readInt()
+        row
+      }
+    }
+  }
+
+  private def ensureTotalSizeIsBelowLimit(sizeOfNextRow: Int): Unit = {
+    totalUncompressedResultSize += sizeOfNextRow
+    maxCollectSize match {
+      case Some(maxSize) => if (totalUncompressedResultSize > maxSize) {
+        val msg = s"Total size of uncompressed results " +
+          s"(${Utils.bytesToString(totalUncompressedResultSize)}) " +
+          s"is bigger than the limit of (${Utils.bytesToString(maxSize)})"
+        logError(msg)
+        throw new SparkException(msg)
```

Does the current behaviour satisfy your original motivation in the PR description? For example, if there are many small partitions, the collect result size could be much bigger, right?

---

That's why we check when encoding each partition (to fail fast if a single partition is already too big) and also we sum up the sizes of all partitions when decoding (to fail if the total collected size of all partitions is too big). So yes, this does satisfy my original motivation.